src/main/scala/strawman/collection/Map.scala

```diff
@@ -67,20 +68,23 @@ trait MapOps[K, +V, +CC[X, Y] <: MapOps[X, Y, CC, _], +C <: MapOps[K, V, CC, C]]
   /** Collects all keys of this map in a set.
     * @return  a set containing all keys of this map.
     */
-  def keySet: Set[K] = new DefaultKeySet
+  def keySet: Set[K] = new KeySet
 
   /** The implementation class of the set returned by `keySet`.
     */
-  protected class DefaultKeySet extends Set[K] with Serializable {
-    def contains(key: K) = MapOps.this.coll.contains(key)
-    def iterator() = keysIterator()
-    override def size = toIterable.size
-    override def foreach[U](f: K => U) = keysIterator() foreach f
+  protected class KeySet extends Set[K] with GenKeySet {
     def iterableFactory: IterableFactory[Set] = Set
-    def empty: Set[K] = iterableFactory.empty
-    def diff(that: Set[K]): Set[K] = fromSpecificIterable(iterableFactory.fromIterable(toIterable).diff(that))
-    protected[this] def fromSpecificIterable(coll: Iterable[K]): Set[K] = iterableFactory.fromIterable(coll)
+    protected[this] def fromSpecificIterable(coll: Iterable[K]): Set[K] = fromIterable(coll)
     protected[this] def newSpecificBuilder(): Builder[K, Set[K]] = iterableFactory.newBuilder()
+    def diff(that: Set[K]): Set[K] = fromSpecificIterable(this).diff(that)
+    def empty: Set[K] = iterableFactory.empty
+  }
+
+  /** A generic trait that is reused by keyset implementations */
+  protected trait GenKeySet extends Serializable { this: Set[K] =>
```

Comment:
Yes, but if you have a nested class you then have to wonder whether the implicitly defined local `this` shadows the explicitly defined outer `this`.  I guess if there isn't a nested class, it's fine; or we can expect people to know the rules regardless.  Maybe this is a personal style I adopted (don't use `this` to name self-types; use `_` if you don't need to name it, or some other name if you need to use it nested) that isn't widely followed.